framework/src/utils/TransitionLayerTools.C

```diff
@@ -0,0 +1,614 @@
+                   const std::vector<unsigned int> node_number_vec,
+                   const std::vector<Real> wt,
+                   const std::vector<Real> index,
+                   const unsigned int boundary_node_num,
+                   const unsigned int i)
+{
+  // First element is trivial
+  weighted_surrogate_index.push_back(0.0);
+  unweighted_surrogate_index.push_back(0.0);
+  for (unsigned int j = 1; j < node_number_vec[i]; j++)
+  {
+    // uniform interval for unweighted index
+    unweighted_surrogate_index.push_back((Real)j / ((Real)node_number_vec[i] - 1.0));
+    // >
+    const auto it_0 =
+        std::upper_bound(index.begin(), index.end(), unweighted_surrogate_index[j - 1]);
+    // >=
+    const auto it_1 = std::lower_bound(index.begin(), index.end(), unweighted_surrogate_index[j]);
+    //
+    const auto it_dist = std::distance(it_0, it_1);
+    //
+    const auto it_start = std::distance(index.begin(), it_0);
+
+    if (it_0 == it_1)
+      weighted_surrogate_index.push_back(weighted_surrogate_index[j - 1] +
+                                         wt[it_start - 1] / ((Real)node_number_vec[i] - 1.0));
+    else
+    {
+      weighted_surrogate_index.push_back(weighted_surrogate_index[j - 1]);
+      weighted_surrogate_index[j] += (*it_0 - unweighted_surrogate_index[j - 1]) * wt[it_start - 1];
+      weighted_surrogate_index[j] +=
+          (unweighted_surrogate_index[j] - *(it_1 - 1)) * wt[it_start + it_dist - 1];
+      for (unsigned int k = 1; k < it_dist; k++)
+        weighted_surrogate_index[j] += wt[it_start + k - 1] / ((Real)boundary_node_num - 1.0);
+    }
+  }
+}
+
+bool
+isXYPlane(const std::vector<Point> vec_pts)
+{
+  for (auto it = vec_pts.begin(); it != vec_pts.end(); it++)
+    if (!MooseUtils::absoluteFuzzyEqual((*it)(2), 0.0))
+      return false;
+  return true;
+}
+
+bool
+needFlip(const std::vector<Point> vec_pts_1, const std::vector<Point> vec_pts_2)
+{
+  const Real th1 =
+      acos((vec_pts_1.back() - vec_pts_1.front()) * (vec_pts_2.front() - vec_pts_1.front()) /
+           (vec_pts_1.back() - vec_pts_1.front()).norm() /
+           (vec_pts_2.front() - vec_pts_1.front()).norm());
+  const Real th2 = acos(
+      (vec_pts_2.back() - vec_pts_1.back()) * (vec_pts_1.front() - vec_pts_1.back()) /
+      (vec_pts_2.back() - vec_pts_1.back()).norm() / (vec_pts_1.front() - vec_pts_1.back()).norm());
+  const Real th3 = acos(
+      (vec_pts_2.front() - vec_pts_2.back()) * (vec_pts_1.back() - vec_pts_2.back()) /
+      (vec_pts_2.front() - vec_pts_2.back()).norm() / (vec_pts_1.back() - vec_pts_2.back()).norm());
+  const Real th4 =
+      acos((vec_pts_1.front() - vec_pts_2.front()) * (vec_pts_2.back() - vec_pts_2.front()) /
+           (vec_pts_1.front() - vec_pts_2.front()).norm() /
+           (vec_pts_2.back() - vec_pts_2.front()).norm());
+  if (MooseUtils::absoluteFuzzyEqual(th1 + th2 + th3 + th4, 2 * M_PI))
+    return false;
+  return true;
+}
+
+bool
+isBoundaryValid(ReplicatedMesh & mesh,
```

Comment:
how big is the overlap with PeripheralRingMeshGenerator's version of this

Reply:
A lot. My plan is to move it here into the namespace as this is in the framework.